docs: update universal doc (#36005)
angular/angular: aio/content/guide/universal.md

To start rendering your app with Universal on your local system, use the following command.

<code-example language="bash">

> npm run dev:ssr

Reviewer: There is in version 9 🤪

> npm run dev:ssr

Reviewer 2: Getting there (but not there just yet)... 😛

</code-example>

> Open a browser and navigate to http://localhost:4200/.

Reviewer: The ssr dev server will be spawned on 4200.

The server-rendered app still launches quickly but the full client app may take seconds to load.

### Useful Scripts

#### npm run dev:ssr

> This command is similar to `ng serve` which offers live reload during the development, the application will run in watch mode and refresh the browser after every change, this is slower to actual `ng serve` command.

Reviewer 3: ```suggestion
This command is similar to `ng serve` which offers live reload during the development. The application will run in watch mode and refresh the browser after every change. This command is slower to actual `ng serve` command.
```
Just changing the commas to periods. Shorter sentences are best! :)

#### npm run build:ssr

> This command is used to build the application for deployment in production mode. This is Shortcut for: ng build --prod && ng run app-name:server:production

Reviewer 3: ```suggestion
Use this command to build the application for deployment in production mode. This is shortcut for `ng build --prod && ng run app-name:server:production`.
```

> Use this command to build the application for deployment in production mode. This is shortcut for `ng build --prod && ng run app-name:server:production`.

Reviewer: The command needs to be updated to `ng build && ng run app-name:server`

#### npm run serve:ssr

> This command is to run the application locally for testing, and should not be executed on remote it runs the node server on your local machine.

Reviewer 3: ```suggestion
Use this command to run the application locally for testing. You should not execute this command on a remote machine as it runs the node server on your local machine.
```

> Use this command to run the application locally for testing. You should not execute this command on a remote machine as it runs the node server on your local machine.

Reviewer 2: Same, here, consider keeping the format consistent. E.g.:

```suggestion
This command starts the server script for serving the application locally with server-side rendering. [...]
```

##
I would also add that this command assumes you have already run `build:ssr` (i.e. that is uses the build artifacts created by `build:ssr`).

##
> You should not execute this command on a remote machine as it runs the node server on your local machine.

I find this sentence a little confusing. It sounds like the command will run the server on your local machine even if you run it on a remote machine :grin:
This is clearly not the case, so if the command runs the server script on the machine that it is invoked on, then this sentence does not explain why not to run it on a remote machine (if your intention is exactly to run the server script on that machine).

I would suggest either removing it or re-phrasing to something like:

```
Note that this command is intended to be used for testing the server-side rendered application locally.
```

> 

Reviewer 2: Would it be worth mentioning the `prerender` script and point to the "Prerendering" guide here?

{@a why-do-it}

> ## Why use server-side rendering?

Reviewer: This should be moved at the top.

The server-rendered app still launches quickly but the full client app may take seconds to load.

### Production Build

> To create production build Angular Universal provides commands to build and test before deployment

Reviewer 4: ```suggestion
To create a production build, use commands provided in Angular Universal to build and test before deployment.
```

> Just run the below command to Run and Test your Universal Application locally:

Reviewer 4: ```suggestion
Run the following command to run and test your universal application locally:
```

<code-example language="bash">

> npm run build:ssr && npm run serve:ssr

Reviewer: Can we explain what each of the above commands do?

## Useful Scripts 
### npm run dev:ssr 
Runs the browser and server applications in the development mode.  The applications will run in watch mode, when doing changes the browser will refresh and the server will be restarted.

### npm run build:ssr 
Builds the browser and server applications in production mode. Shortcut for: `ng build --prod && ng run app-name:server:production `

### npm run serve:ssr 
Launches the Node.js server and runs the server application.  

Note: phases can be improved, just a short example.

</code-example>

### Prerendering

> With release of Angular 9 it is possible to prerender the static pages as part of the build, it uses the guess parser to prerender the static pages.

Reviewer: This is actually not a feature of `Angular 9`, but rather of `Angular Universal version 9`.

> With release of Angular 9 it is possible to prerender the static pages as part of the build, it uses the guess parser to prerender the static pages.

Reviewer 4: ```suggestion
As of of Angular Universal version 9, you can now pre-render the static pages as part of the build using the guess parser.
```

> You can run the below command to prerender static page:

Reviewer 4: ```suggestion
To pre-render a static page, run the following command:
```

<code-example language="bash">
npm run prerender
</code-example>

> ### Prerendering Dynamic Routes

Reviewer 4: ```suggestion
### Pre-rendering Dynamic Routes
```

> Prerendering dyanmic routes is possible for example for a route like `product/:id` where `id` is dynmically provided.

Reviewer 4: ```suggestion
You can pre-render a dynamic route, such as `product/:id` in which `id` is provided dynamically.
```

> To achieve this you can choose one from below 2 options:

Reviewer 4: ```suggestion
Choose one of the following methods:
```

> #### Providing manual routes

Reviewer: `Providing manual routes` -> `Providing additional routes`

Clarify when the `--routes` can be used, even outside of dynamic routes.

> While running the prerender command you can pass the values like below:

Reviewer 4: ```suggestion
While running the `prerender` command, pass values as follows:
```

<code-example language="bash">
ng run &lt;app-name&gt;:prerender --routes /product/1 --routes /product/2
</code-example>

> #### Providing all routes via file

Reviewer: `Providing all routes via file`, the title here is confusion. Because when using the `routeFile` option, you don't need to provide all the routes. Since routes will be merged with the once provided via the `routes` option and those which are guessed via the parser.

I think it may be best to explain each builder option. Example, from the doc, I don't understand how can I opt out of route guessing etc..., as while these options are used for dynamic routes, they can be used for other use cases.

So my suggestion would be to explain the options, and than have a separate sub section for dynamic routes.

> #### Providing all routes via file

Reviewer 4: ```suggestion
#### Providing all routes in a file
```

> The above option is good where we have limited number of parameters, but what in case we have many records and in database and we want to generate static pages, this is where the second option fits in.

Reviewer 4: Move this up to the previous section at the end.

```suggestion
Providing manual routes is the best option with a limited number of parameters.
```

> The above option is good where we have limited number of parameters, but what in case we have many records and in database and we want to generate static pages, this is where the second option fits in.

Reviewer 4: ```suggestion
Provide routes in a file to generate static pages if you have many more parameters, such as multiple records in a database.
```

> You can pass a `.txt` file with list of `id` with `--routeFile` option with `prerender` command.

Reviewer 4: ```suggestion
To pass a `.txt` file with a list of IDs, use the `--routeFile` option with the `prerender` command.
```

> You can fetch list of ids using nodejs and create a `routes.txt` file with values like:

Reviewer 4: ```suggestion
You can fetch list of IDs using Node.js, and create a `routes.txt` file with values such as the following:
```

/products/1
/products/555

> and once txt file is ready, run the below command to prerender the static files with dynamic values:

Reviewer 4: ```suggestion
When your `.txt` file is ready, run the following command to pre-render the static files with dynamic values:
```

Here, "request URL" refers to the URL of the request as a response to which the application is being rendered on the server.
For example, if the client requested `https://my-server.com/dashboard` and you are rendering the application on the server to respond to that request, `options.url` should be set to `https://my-server.com/dashboard`.

Now, on every HTTP request made as part of rendering the app on the server, Angular can correctly resolve the request URL to an absolute URL, using the provided `options.url`.
### Useful scripts

-  `npm run dev:ssr`

> This command is similar to [`ng serve`](cli/serve), which offers live reload during development, but uses server-side rendering. The application will run in watch mode and refresh the browser after every change. This command is slower than the actual `ng serve` command.

Reviewer 2: This needs to be indented to by considered part of the bullet-point:

```suggestion
  This command is similar to [`ng serve`](cli/serve), which offers live reload during development, but uses server-side rendering. The application will run in watch mode and refresh the browser after every change.
  This command is slower than the actual `ng serve` command.
```

Same for the other bullet-points below.

- `ng build && ng run app-name:server`

This command builds both the server script and the application in production mode. Use this command when you want to build the project for deployment.

- `npm run serve:ssr`

This command starts the server script for serving the application locally with server-side rendering. It uses the build artifacts created by `ng run build:ssr`, so make sure you have run that command as well.

> 

Reviewer 2: As mentioned in https://github.com/angular/angular/pull/36005#discussion_r637932110, I think it is valuable to add the following note:

```suggestion

Note that `serve:ssr` is not intended to be used to serve your application in production, but only for testing the server-side rendered application locally.

```
